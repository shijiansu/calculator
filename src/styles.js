import { StyleSheet, Dimensions } from 'react-native';
import { colors, iphoneXHeight, iphoneXWidth } from './constant';

let { width, height } = Dimensions.get('window');

export const changeValue = (newWidth, newHeight) => {
  width = newWidth;
  height = newHeight;
};

export const setValue = value => {
  const ratio = (height * width) / (iphoneXHeight * iphoneXWidth);
  return ratio >= 1 ? value : value * ratio;
};

export const setYAxisValue = value => {
  const ratio = height / iphoneXHeight;
  return ratio > 1 ? value : value * ratio;
};

export const setXAxisValue = value => {
  const ratio = width / iphoneXWidth;
  return ratio >= 1 ? value : value * ratio;
};

const button = {
  margin: setValue(7),
  width: setValue(70),
  height: setValue(70),
  justifyContent: 'center',
  alignItems: 'center',
  backgroundColor: colors.GREY,
  borderRadius: setValue(70) / 2,
  maxHeight: setYAxisValue(82),
};

const buttonText = { fontSize: setXAxisValue(28), fontWeight: '500' };

export default StyleSheet.create({
  buttonDefault: { ...button },
  buttonType1: {
    ...button,
    backgroundColor: colors.ORANGE,
  },
  buttonTextDefault: { ...buttonText },
  buttonTextType1: {
    ...buttonText,
    fontSize: setXAxisValue(30),
    maxHeight: setYAxisValue(42),
  },
  screenText: {
    fontSize: setXAxisValue(60),
    fontWeight: '400',
    paddingHorizontal: setXAxisValue(15),
  },
  numberPad: { flex: 1, flexDirection: 'row' },
  screen: {
    justifyContent: 'flex-end',
    alignItems: 'flex-end',
    borderWidth: StyleSheet.hairlineWidth,
    marginTop: setYAxisValue(40),
    marginBottom: setYAxisValue(30),
    marginHorizontal: setXAxisValue(10),
  },
  container: { flex: 1, padding: setValue(10) },
  safeView: { flex: 1, alignItems: 'center' },
  headerView: {
    height: setYAxisValue(40),
    width: '100%',
    borderBottomColor: colors.BLACK,
    borderBottomWidth: StyleSheet.hairlineWidth,
    alignItems: 'center',
    justifyContent: 'center',
    flexDirection: 'row',
  },
  resetButton: {
    alignSelf: 'center',
    position: 'absolute',
    right: setXAxisValue(10),
  },
  headerTitle: {
    fontSize: setXAxisValue(18),
    fontWeight: 'bold',
  },
  resetText: {
    fontSize: setXAxisValue(14),
  },
});
