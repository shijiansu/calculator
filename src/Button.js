import React from 'react';
import { Text, TouchableOpacity } from 'react-native';
import styles from './styles';

const Button = ({ text, type, onPress }) => {
  const containerStyle = type === 0 ? styles.buttonDefault : styles.buttonType1;
  const textStyle = type === 0 ? styles.buttonTextDefault : styles.buttonTextType1;
  return (
    <TouchableOpacity style={containerStyle} onPress={onPress}>
      <Text style={textStyle}>{text}</Text>
    </TouchableOpacity>
  );
};

export default React.memo(Button);
