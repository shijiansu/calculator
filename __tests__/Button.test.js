import React from 'react';
import { shallow } from 'enzyme';
import Button from '../src/Button';
import styles from '../src/styles';

describe('A button with type 0', () => {
  let component;
  let onPress;

  beforeEach(() => {
    onPress = jest.fn();
    component = shallow(<Button text="0" type={0} onPress={onPress} />);
  });

  it('should renders correctly', () => {
    expect(component).toMatchSnapshot();
  });

  it('should have correct text', () => {
    expect(
      component
        .find('Text')
        .children()
        .text()
    ).toEqual('0');
  });

  it('should call onPress', () => {
    component
      .find('TouchableOpacity')
      .at(0)
      .simulate('press');
    expect(onPress).toHaveBeenCalled();
  });

  it('should have button style of type 0', () => {
    expect(
      component
        .find('TouchableOpacity')
        .at(0)
        .props().style
    ).toEqual(styles.buttonDefault);
    expect(
      component
        .find('Text')
        .at(0)
        .props().style
    ).toEqual(styles.buttonTextDefault);
  });
});

describe('A button with type 1', () => {
  let component;
  let onPress;

  beforeEach(() => {
    onPress = jest.fn();
    component = shallow(<Button text="1" type={1} onPress={onPress} />);
  });

  it('should renders correctly', () => {
    expect(component).toMatchSnapshot();
  });

  it('should have correct text', () => {
    expect(
      component
        .find('Text')
        .children()
        .text()
    ).toEqual('1');
  });

  it('should call onPress', () => {
    component
      .find('TouchableOpacity')
      .at(0)
      .simulate('press');
    expect(onPress).toHaveBeenCalled();
  });

  it('should have button style of type 1', () => {
    expect(
      component
        .find('TouchableOpacity')
        .at(0)
        .props().style
    ).toEqual(styles.buttonType1);
    expect(
      component
        .find('Text')
        .at(0)
        .props().style
    ).toEqual(styles.buttonTextType1);
  });
});
