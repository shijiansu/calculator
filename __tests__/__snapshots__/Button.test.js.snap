// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A button with type 0 should renders correctly 1`] = `ShallowWrapper {}`;

exports[`A button with type 1 should renders correctly 1`] = `ShallowWrapper {}`;
